### PR TITLE
Vendor rust dependencies

### LINKED ISSUE #61741: Rust does not work without Internet
If you do `cmake` without internet access, it will print something like `Couldn't resolve host: index.crates.io` which is ridiculous.
Workaround: `cmake -D ENABLE_RUST=0`

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Vendor in rust dependencies.

### Documentation entry for user-facing changes

Rust dependencies are no longer retrieved from the internet and they live in a submodule.

> [!IMPORTANT]  
> This PR also updates the internal rust toolchain to **nightly-2024-04-01**. As the vendor dependencies for the sanitizer builds are specific a rustc version, using other versions might not be supported (due to missing deps in the submodule).

Closes https://github.com/ClickHouse/ClickHouse/issues/61741

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by Algunenano** [comment]
Ok it seems it's finally building in all scenarios. I needed to vendor not only dependencies but also the crate repositories themselves. Without them you still need a connection.

Some things need to be changed though:
* It always rebuild the rust libraries even with no changes to them. :suboptimal:
* The vendor directories are pretty big. This is because you can't ignore certain architectures/OSx while vendoring and windows has extremely big binary deps. No easy solution that I could see, but it might be possible to ignore those nested deps with some patch or something like that.
* The vendor deps might be duplicated since they've been created for each of the 2 projects. I'll see if I can merge them.

After everything is ready, it would be best to clean up the history of the submodule before merging to remove all the big files from history and keep only the necessary ones.

**Comment by Algunenano** [comment]
After way more hours that I'm ok accepting, trying I don't know how many things to solve the recompilation problem, I found out the problem with rebuilding comes from having multiple corrosion projects clashing. The solution is to do things properly and create, and use, a rust workspace. This fixes both the first and the third point.

Only thing pending is to see if there is a way to remove the windows deps from the vendor submodule so it's not that big.

**Comment by Algunenano** [comment]
A couple of problems with the build:
* FreeBSD: The new rust toolchain used new functions that aren't available in our sysroot. Support for rust and FreeBSD was already partially disabled and I've disabled it completely for now. A possible solution would be to update the sysroot to something newer, but that might mean losing compatibility with older FreeBSD (maybe ok?).
* MUSL: The new rust toolchain uses a recent libc function (`pthread_getname_np`) that wasn't available in our musl sysroot. I'm updating the sysroot here: https://github.com/ClickHouse/sysroot/pull/22

**Comment by Algunenano** [comment]
Everything builds and tests pass, so I'm updating the submodules to their final status

**Comment by Algunenano** [comment]
This is ready and has a green CI. Only missing review and merge with master to fix conflicts

**Review by alexey-milovidov — APPROVED** [review]

File: rust/VENDOR.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+As we have multiple projects we use a workspace to manage them (it's way simpler and leads to less issues). In order
```
Comment (Felixoid) [inline rust/VENDOR.md:1]: A very cool thing is done here!

Does adding a script into https://github.com/ClickHouse/rust_vendor to update everything as a simple command make sense?

Comment (Algunenano) [reply rust/VENDOR.md:1]: Maybe.

I generated the commands matching exactly how `std` is organised and our current `Cargo.lock` so they might not be universal. We can wait to execute them a few times in the future (understanding the commands) before making it automatic.

**Pull request merged**